source/ice_api.c

```diff
@@ -0,0 +1,1676 @@
+        pIceAgent->computeCRC32 = computeCRC32Function;
+        pIceAgent->computeHMAC = computeHMACFunction;
+
+        /* This field is required as an attribute during creation of STUN packet. */
+        pIceAgent->tieBreaker = pIceAgent->computeRandom();
+    }
+
+    return retStatus;
+}
+/*------------------------------------------------------------------------------------------------------------------*/
+
+/* Ice_AddHostCandidate - The application calls this API for adding host candidate. */
+
+IceResult_t Ice_AddHostCandidate( const IceIPAddress_t ipAddr,
+                                  IceAgent_t * pIceAgent,
+                                  IceCandidate_t ** ppCandidate )
+{
+    IceResult_t retStatus = ICE_RESULT_OK;
+    IceCandidate_t * pIceCandidate;
+    int localCandidateCount;
+
+    if( ( pIceAgent == NULL ) ||
+        ( ppCandidate == NULL ) )
+    {
+        retStatus = ICE_RESULT_BAD_PARAM;
+    }
+
+    if( retStatus == ICE_RESULT_OK )
+    {
+        localCandidateCount = Ice_GetValidLocalCandidateCount( pIceAgent );
+    }
+
+    if( localCandidateCount >= ICE_MAX_LOCAL_CANDIDATE_COUNT )
+    {
+        retStatus = ICE_RESULT_MAX_CANDIDATE_THRESHOLD;
+    }
+    else
+    {
+        pIceCandidate = &pIceAgent->localCandidates[ localCandidateCount ];
+    }
+
+    if( retStatus == ICE_RESULT_OK )
+    {
+        pIceCandidate->isRemote = 0;
+        memcpy( &pIceCandidate->ipAddress,
+                &ipAddr,
+                sizeof( IceIPAddress_t ) );
+        pIceCandidate->iceCandidateType = ICE_CANDIDATE_TYPE_HOST;
+        pIceCandidate->state = ICE_CANDIDATE_STATE_VALID;
+        pIceCandidate->priority = Ice_ComputeCandidatePriority( pIceCandidate );
+    }
+
+    if( retStatus == ICE_RESULT_OK )
+    {
+        *ppCandidate = pIceCandidate;
+    }
+
+    return retStatus;
+}
+/*------------------------------------------------------------------------------------------------------------------*/
+
+/* Ice_AddSrflxCandidate - The application calls this API for adding Server Reflexive candidate. */
+
+IceResult_t Ice_AddSrflxCandidate( const IceIPAddress_t ipAddr,
+                                   IceAgent_t * pIceAgent,
+                                   IceCandidate_t ** ppCandidate,
+                                   uint8_t * pTransactionIdBuffer,
+                                   uint8_t ** ppSendStunMessageBuffer,
+                                   uint32_t * pSendStunMessageBufferLength )
+{
+    IceResult_t retStatus = ICE_RESULT_OK;
+    IceCandidate_t * pIceCandidate;
+    int localCandidateCount;
+
+    if( ( pIceAgent == NULL ) ||
+        ( ppCandidate == NULL ) ||
+        ( ppSendStunMessageBuffer == NULL ) ||
+        ( pSendStunMessageBufferLength == NULL ) ||
+        ( pTransactionIdBuffer == NULL ) )
+    {
+        retStatus = ICE_RESULT_BAD_PARAM;
+    }
+
+    localCandidateCount = Ice_GetValidLocalCandidateCount( pIceAgent );
```

Comment:
It should be inside if( result == ICE_RESULT_OK), if pIceAgent  is NULL, this could be a NULL access.

Reply:
Updated with `else`